pkg/workspace/assert.go

```diff
@@ -0,0 +1,167 @@
+	}
+	if len(annotation.Args) == 0 {
+		return nil, fmt.Errorf("Expected @%v to have validation 2-tuple as argument(s), but found no arguments (by %v)", AnnotationAssertValidate, validationPosition.AsCompactString())
+	}
+	for _, arg := range annotation.Args {
+		validationTuple, ok := arg.(starlark.Tuple)
+		if !ok {
+			return nil, fmt.Errorf("Expected @%v to have validation 2-tuple as argument(s), but found: %v (by %v)", AnnotationAssertValidate, arg.String(), validationPosition.AsCompactString())
+		}
+		if len(validationTuple) != 2 {
+			return nil, fmt.Errorf("Expected @%v 2-tuple, but found tuple with length %v (by %v)", AnnotationAssertValidate, len(validationTuple), validationPosition.AsCompactString())
+		}
+		message, ok := validationTuple[0].(starlark.String)
+		if !ok {
+			return nil, fmt.Errorf("Expected @%v to have descriptive string as the first item in the 2-tuple, but found type: %v (by %v)", AnnotationAssertValidate, validationTuple[0].Type(), validationPosition.AsCompactString())
+		}
+		lambda, ok := validationTuple[1].(starlark.Callable)
+		if !ok {
+			return nil, fmt.Errorf("Expected @%v to have a validation function as the second item in the 2-tuple, but found type: %v (by %v)", AnnotationAssertValidate, validationTuple[1].Type(), validationPosition.AsCompactString())
+		}
+		validations = append(validations, Validation{
+			thread:   thread,
+			msg:      message.String(),
+			f:        lambda,
+			position: validationPosition,
+		})
+	}
+
+	return validations, nil
+}
+
+// Validation is an assertion from an @assert/validate annotation.
+// One @assert/validate annotation can have multiple Validations.
+type Validation struct {
+	thread   *starlark.Thread
+	msg      string
+	f        starlark.Callable
+	position *filepos.Position
+}
+
+// Validate runs the function from the Validation with the node's value as arguments.
+//
+// Returns an error if the function returns an error, or if the resulting value from the function is falsy.
+// Otherwise, returns nil.
+func (v Validation) Validate(node yamlmeta.Node) error {
+	var nodeValue starlark.Value
+	switch values := node.GetValues(); len(values) {
+	case 0: // run validation with no arguments
+	case 1:
+		nodeValue = yamltemplate.NewGoValueWithYAML(values[0]).AsStarlarkValue()
+	default:
+		nodeValue = yamltemplate.NewGoValueWithYAML(values).AsStarlarkValue()
+	}
+
+	result, err := starlark.Call(v.thread, v.f, starlark.Tuple{nodeValue}, []starlark.Tuple{})
+	if err != nil {
+		return fmt.Errorf("%v requires a valid value: %v; %v (by %v)", node.GetPosition().AsCompactString(), v.msg, err.Error(), v.position.AsCompactString())
+	}
+	// From proposal: "if all rules pass (either returns True or None), then the value is valid."
+	// I believe a None result would return an error
+	if !result.Truth() {
```

Comment:
🤔 

I really dig the simplicity of "Falsey is faliure, Truthy is success."

_However_, it's really common for assertion functions to actually return nil/null/None upon success.

For example:

https://github.com/vmware-tanzu/carvel-ytt/blob/b50fd584133ce57a147c9fa050bd4acea684a259/pkg/yttlibrary/assert.go#L75-L86

Reply:
Great point, I'm not sure where I got the Falsey as failure for assertions from, after reviewing the proposal, I see here what the desired behavior is:
- [from proposal](https://github.com/vmware-tanzu/carvel/blob/aa19de3af4626f70765b8543bea60ca4ff5ca9c2/proposals/ytt/004-schema-validation/README.md#assertvalidate)
- if all rules pass (either returns True or None), then the value is valid.
- if a rule returns False (not None) or assert.fail()s, then the valid is invalid.

Reply:
Yup yup...

and a fine-tuning:

> if a rule returns Falsey (except None) or assert.fail()s, then the valid is invalid.